crates/core/tedge/src/cli/mqtt/mod.rs

```diff
@@ -13,29 +11,51 @@ mod subscribe;
 
 const MAX_PACKET_SIZE: usize = 268435455; // 256 MB
 
-fn disconnect_if_interrupted(client: Client, timeout: Option<Duration>) -> Arc<AtomicBool> {
-    let interrupted = Arc::new(AtomicBool::new(false));
-    for signal in signal_hook::consts::TERM_SIGNALS {
-        let client = client.clone();
-        let interrupted = interrupted.clone();
-        unsafe {
-            let _ = signal_hook::low_level::register(*signal, move || {
-                eprintln!("INFO: Interrupted");
-                interrupted.store(true, Ordering::Relaxed);
-                let _ = client.disconnect();
+#[derive(Copy, Clone, Debug)]
+enum Interruption {
+    Timeout,
+    Interrupted,
+    Error,
+}
+
+struct TermSignals(watch::Receiver<Option<Interruption>>);
+
+impl TermSignals {
+    fn new(timeout: Option<Duration>) -> TermSignals {
+        let (tx, rx) = watch::channel(None);
+        for signal in signal_hook::consts::TERM_SIGNALS {
```

Comment:
I temporary stick with the blocking API for signals. Once, the `tedge` command ready to run under `tokio::main` if will move an async API for that.

Reply:
Done: https://github.com/thin-edge/thin-edge.io/pull/3466/commits/d66b2d360c8a09faf076af4bbf8e7cd3959d9e63